src/Orleans.Runtime/Timers/GrainTimer.cs

```diff
@@ -124,10 +133,34 @@ private async Task ForwardToAsyncCallback(object state)
             {
                 previousTickTime = DateTime.UtcNow;
                 currentlyExecutingTickTask = null;
+
                 // if this is not a repeating timer, then we can
                 // dispose of the timer.
                 if (timerFrequency == Constants.INFINITE_TIMESPAN)
-                    DisposeTimer();                
+                {
+                    DisposeTimer();
+                }
+            }
+        }
+
+        private async Task InvokeAsyncCallback(object state)
+        {
+            // This is called under a lock, so ensure that the method yields before invoking a callback
+            // which could take a different lock and potentially cause a deadlock.
+            await Task.Yield();
```

Comment:
This is the crux of the fix: under the lock, we forward the call to this method, which immediately yields, releasing the lock before the callback is invoked. This way, the callback will not be invoked under the lock and the potential for a deadlock is eliminated.